include-what-you-use.1

```diff
@@ -113,10 +123,17 @@ is already visible in the file's transitive includes.
 Set verbosity. At the highest level, this will dump the AST of the source file
 and explain all decisions.
 .SH EXIT STATUS
+By default
 .B include-what-you-use
-always returns with a nonzero status code to make usage with
-.BR make (1)
-feasible.
+exits with zero exit code unless there's a critical error, but
+.B \-\-fail or
+.B \-\-exit_code
+can be used to customize the exit-code in depending on invoker expectations.
+For example, it can be useful to always return a failure exit code to
+.B make(1)
+to force reruns. For example:
+.B make -k CXX=include-what-you-use CXXFLAGS="-Xiwyu --fail"
+
```

Comment:
Nice, I hadn't seen the parallel between `-Werror` and `--exit_code`. I wonder if there's some spelling of `--exit_code` that brings that similarity to light.

Reply:
The above comment is incorrect, since I was a bit confused by the description of `--fail` above.

Reply:
It's a tricky matrix;

* If something is critically wrong, e.g. bad args or parse errors, we exit with 1
* If `--exit_code` is provided, the exit code signals that there are IWYU warnings. Compare `-Werror` or `git diff --exit-code` or `grep -q`
* `--exit_code` defaults to `1`, but you can specify a particular exit code (e.g. `2` or `99`), in case users want to take different action depending on parse errors vs IWYU warnings
* `--fail` exits with error whether there are IWYU warnings or not. For use with `make -k`.
* The particular exit code of `--fail` can be customized using `--exit_code=N`

So `--exit_code` serves two purposes; one is to say that IWYU should exit with error if there are warnings, the other is to say which exit code it should use if it exits with error. `--fail` just forces it to exit with error.

It's a little confusing, but I haven't been able to figure out anything that's less confusing without becoming overly verbose :-)

Reply:
Ahh, now I understand.  What if, `--fail` accepts an optional value, indicating its exit status?  That way, each option is independent of each other.  And you could have `--error` (or `--Werror`, or whatever) for transforming warnings into errors, and it could also have an optional value, indicating its exit status.  Both of them should default to one if no argument is provided, or 0 if the option is not specified.

Reply:
I was doing some laundry and had some time to reflect. This makes sense to me:

* `--error[=N]` -- exit with N (defaults to 1) if there are IWYU warnings
* `--error_always[=N]` -- exit with N (defaults to 1) no matter what

There's still a chance that people might pass both, e.g. `-Xiwyu --error=10 -Xiwyu --error_always`, but it feels intuitive that `--error_always` takes precedence here. We could even warn if both are provided.

Reply:
Your `--error` and `--error_always` are exactly the same as my `--error` and `--fail` suggestion above, so I like it.  In case a user specifies both, there are 2 things that make sense:  the one I suggested, or warn the user.

Reply:
> Your --error and --error_always are exactly the same as my --error and --fail suggestion above, so I like it. I'm not sure
> which name I prefer.

Oops, I had a cached page when I typed up my last response, so I hadn't seen your post. Nice.

I like that they are both called `--error.*` to indicate that they're related. Or even mutually exclusive.

I was not planning on letting them interplay like you suggested, but maybe that would be nice. It's easy to change should it ever become necessary.

> BTW, I think it's a bit weird the iwyu convention of --xxx_yyy for options instead of the typical --xxx-yyy :)

I agree, but it would be even weirder to start mixing `--xx_yy` and `--xx-zz` at this point. Maybe there's a future project to set everything straight :-)

Reply:
Yes, if you plan to not allow mixing them, your naming is much more explicit and clear.  I like it.

And yes, consistency takes precedence over conventions here :-)

Reply:
Updated.